sbom-utility-scripts/scripts/merge-cachi2-sboms-script/merge_cachi2_sboms.py

```diff
@@ -148,6 +299,199 @@ def _merge_tools_metadata(syft_sbom: dict[Any, Any], cachi2_sbom: dict[Any, Any]
+    root_element2, map2, inverse_map2 = map_relationships(relationships2)
+    package_ids = [package["SPDXID"] for package in packages]
+
+    root_package1 = calculate_root_package(root_element1, map1, inverse_map1)
+    root_package2 = calculate_root_package(root_element2, map2, inverse_map2)
+
+    for relation in relationships2:
+        _relation = relation.copy()
+
+        # If relations is Root decribes middle element, skip it
+        if (
+            _relation["relatedSpdxElement"] == root_package2
+            and _relation["spdxElementId"] == root_element2
+            and _relation["relationshipType"] == "DESCRIBES"
+        ):
+            continue
+        # if spdxElementId is root_element2, replace it with root_element1
+        # if not and relatedSpdxElement is root_element2, replace it with root_element1
+        if _relation["spdxElementId"] == root_element2:
+            _relation["spdxElementId"] = root_element1
+        elif relation["relatedSpdxElement"] == root_element2:
+            _relation["relatedSpdxElement"] = root_element1
+        if _relation["spdxElementId"] == root_package2:
+            _relation["spdxElementId"] = root_package1
+        if _relation["relatedSpdxElement"] == root_package2:
+            _relation["relatedSpdxElement"] = root_package1
+
+        # include only relations to packages which exists in merged packages.
+        if _relation["relatedSpdxElement"] in package_ids:
+            relationships.append(_relation)
+        elif _relation["spdxElementId"] in package_ids:
+            relationships.append(_relation)
+
+    for relation in relationships1:
+        _relation = relation.copy()
+        if relation["relatedSpdxElement"] in package_ids:
+            relationships.append(_relation)
+    return relationships
+
+
+def merge_packages(syft_sbom: dict, cachi2_sbom: dict) -> dict:
+    """Merge Cachi2 packages into the Syft SBOM while removing duplicates."""
+
+    def get_package_key(pkg):
+        return json.dumps(sorted(set(_unique_keys_syft_spdx(pkg))), separators=(",", ":"))
+
+    is_duplicate_package = _get_syft_package_filter(cachi2_sbom["packages"])
+    cachi2_packages_map = {get_package_key(p): p for p in cachi2_sbom["packages"]}
+
+    filtered_packages = []
+    for p in syft_sbom.get("packages", []):
+        if is_duplicate_package(p):
+            if get_package_key(p) in cachi2_packages_map:
+                cpackage = cachi2_packages_map[get_package_key(p)]
+                cpackage["externalRefs"] = sorted(
+                    merge_external_refs(cpackage.get("externalRefs", []), p.get("externalRefs", [])),
+                    key=lambda x: (
+                        x["referenceCategory"],
+                        x["referenceType"],
+                        x["referenceLocator"],
+                    ),
+                )
+                cpackage["annotations"] = merge_annotations(cpackage.get("annotations", []), p.get("annotations", []))
```

Comment:
for cyclonedx if the component is duplicated it's based on purl. That means the component is duplicate for sure. With SPDX considering possibility of multiple purls, is_duplicate_package returns True even on partial match of purls. Therefore considered removal of duplicated packages as possible  information loss and that's why attributes of the duplicate package are merged with existing one instead of discarding whole package